src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +616,303 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
 }
 
 
+/*
+ * RecursivelyPlanRecurringTupleOuterJoins descends into a join tree and
+ * recursively plans all non-recurring (i.e., distributed) rels that that
+ * participate in an outer join expression together with a recurring rel,
+ * such as <distributed> in "<recurring> LEFT JOIN <distributed>", i.e.,
+ * where the recurring rel causes returning recurring tuples from the worker
+ * nodes.
+ */
+static void
+RecursivelyPlanRecurringTupleOuterJoins(Node *node, Query *query,
+										RecursivePlanningContext *recursivePlanningContext)
+{
+	if (node == NULL)
+	{
+		return;
+	}
+	else if (IsA(node, RangeTblRef))
+	{
+		return;
+	}
+	else if (IsA(node, FromExpr))
+	{
+		FromExpr *fromExpr = (FromExpr *) node;
+		ListCell *fromExprCell;
+
+		/* search for join trees in each FROM element */
+		foreach(fromExprCell, fromExpr->fromlist)
+		{
+			Node *fromElement = (Node *) lfirst(fromExprCell);
+
+			RecursivelyPlanRecurringTupleOuterJoins(fromElement, query,
+													recursivePlanningContext);
+		}
+
+		return;
+	}
+	else if (IsA(node, JoinExpr))
+	{
+		JoinExpr *joinExpr = (JoinExpr *) node;
+
+		Node *leftNode = joinExpr->larg;
+		Node *rightNode = joinExpr->rarg;
+
+		/* there may be recursively plannable outer joins deeper in the join tree */
+		RecursivelyPlanRecurringTupleOuterJoins(joinExpr->larg, query,
+												recursivePlanningContext);
+		RecursivelyPlanRecurringTupleOuterJoins(joinExpr->rarg, query,
+												recursivePlanningContext);
+		switch (joinExpr->jointype)
+		{
```

Comment:
I think one point that makes this PR harder to follow is that:  We get into RecursivelyPlanRecurringTupleOuterJoins -- which recursively traverses the node -- and then few lines later, we re-travers the same node(s) via IsJoinNodeRecurring.

Is there a simple way to teach RecursivelyPlanRecurringTupleOuterJoins whether the node has/is IsJoinNodeRecurring?

I don't want to complicate the code, but at this point, the flow is a little complicated for me. We go deep down, process main level, go deep down again.

If there is no easy code change, would be good to document why/what we are doing here